lib/Elevate/Components/CCS.pm

```diff
@@ -0,0 +1,388 @@
+package Elevate::Components::CCS;
+
+=encoding utf-8
+
+=head1 NAME
+
+Elevate::Components::CCS
+
+pre_leapp: Export CCS data to a root owned backup directory and remove CCS
+           package
+
+post_leapp: Install CCS package and import the backups taken during pre_leapp
+
+=cut
+
+use cPstrict;
+
+use Try::Tiny;
+
+use File::Path ();
+use File::Copy ();
+
+use Cpanel::Autodie       ();
+use Cpanel::Config::Users ();
+use Cpanel::JSON          ();
+use Cpanel::Pkgr          ();
+
+use Elevate::Notify    ();
+use Elevate::StageFile ();
+
+use Log::Log4perl qw(:easy);
+
+use parent qw{Elevate::Components::Base};
+
+use constant CCS_PACKAGE        => 'cpanel-ccs-calendarserver';
+use constant EXPORT_DIR         => '/var/cpanel/elevate_ccs_export';
+use constant CCS_RESTART_SCRIPT => '/usr/local/cpanel/scripts/restartsrv_cpanel_ccs';
+use constant TASK_QUEUE_SCRIPT  => '/usr/local/cpanel/bin/servers_queue';
+
+use constant DUMP_TYPES => (
+    calendars => 'ics',
+    contacts  => 'vcard',
+);
+
+sub pre_leapp ($self) {
+    my $ccs_installed = Cpanel::Pkgr::is_installed(CCS_PACKAGE);
+    Elevate::StageFile::update_stage_file( { ccs_installed => $ccs_installed } );
+    return unless $ccs_installed;
+
+    $self->_load_ccs_modules();
+
+    $self->run_once('export_ccs_data');
+    $self->remove_ccs_and_dependencies();
+
+    # Removing the PKG will leave this directory in place
+    # This results in PostGreSQL/CCS failing to start after leapp completes
+    $self->clean_up_pkg_cruft();
+
+    return;
+}
+
+sub clean_up_pkg_cruft ($self) {
+    $self->move_pgsql_directory();
+    $self->remove_cpanel_ccs_home_directory();
+    return;
+}
+
+=head1 remove_cpanel_ccs_home_directory
+
+Removing the package removes the `cpanel-ccs` user, but leaves behind
+the data in the home directory.  This results in the service failing
+to start after the elevation has completed
+
+=cut
+
+sub remove_cpanel_ccs_home_directory ($self) {
+    File::Path::remove_tree('/opt/cpanel-ccs') if -d '/opt/cpanel-ccs';
+    return;
+}
+
+=head1 move_pgsql_directory
+
+Removing the PKG will leave this directory in place
+This results in PostGreSQL/CCS failing to start after leapp completes
+
+=cut
+
+sub move_pgsql_directory ($self) {
+    File::Copy::move( '/var/lib/pgsql', '/var/lib/pgsql_pre_elevate' ) if -d '/var/lib/pgsql';
```

Comment:
What if the `/var/lib/pgsql_pre_elevate` directory already exists?

Reply:
Good call, I'll go ahead and make sure it does not exist prior to moving the datadir out of the way